### PR TITLE
Add `$addToSet` update operator

### LINKED ISSUE #73: Implement array update operator `$addToSet`
Beyond "simple" array update operators -- `$push` and `$pop`, see #72 -- there is another that are relatively straight-forward to add:

* `$addToSet`: similar to `$push` but prevents addition of duplicates

(earlier issue also covered `$pullAll` but removed as out-of-initial-scope)

### PR DESCRIPTION
**Pull request by tatu-at-datastax**
**What this PR does**:

Adds implementation of `$addToSet` update operator -- including `$each` modifier.

Includes new helper class that implements order-sensitive (for JSON Objects) equality comparison, needed for update operators (to solve #190)

**Which issue(s) this PR fixes**:
Fixes #73

**Checklist**
- [ ] Changes manually tested
- [x] Automated Tests added/updated
- [ ] Documentation added/updated
- [x] CLA Signed: [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

**Review by vkarpov15 — COMMENTED** [review]
LGTM so far. Just a note that it is important to support `$each` for `$addToSet`, equivalent to `$push`'s `$each` modifier. [Mongoose wraps `$addToSet` in `$each` when using Mongoose's `addToSet()` array helper](https://github.com/Automattic/mongoose/blob/e69b90ca3cb4259058358606a7bf5e9b26eeb229/lib/model.js#L696-L698).

**Comment by tatu-at-datastax** [comment]
@vkarpov15 Good point -- I forgot to add a note that `$each` is/will be supported; changed description to state that.

Will need to add more testing & handling of order-sensitive equality of Object (sub-doc) values; otherwise complete.

File: src/test/java/io/stargate/sgv2/jsonapi/api/v1/FindAndUpdateIntegrationTest.java [hunk 1]
```diff
@@ -1520,6 +1520,123 @@ public void findByColumnAndInc() {
...
+                      }
+                      """);
+      String json =
+          """
+                      {
+                        "updateOne": {
+                          "filter" : {"_id" : "update_doc_add_to_set_each"},
+                          "update" : {
+                              "$addToSet" : {
+                                 "nested.array": { "$each" : [ 1, 3, 4 ] },
```
Comment (ivansenic) [inline src/test/java/io/stargate/sgv2/jsonapi/api/v1/FindAndUpdateIntegrationTest.java:1595]: woud be great if this would also be in some swagger examples, maybe we should think about this in future

File: src/test/java/io/stargate/sgv2/jsonapi/api/v1/FindAndUpdateIntegrationTest.java [hunk 2]
```diff
@@ -1520,6 +1520,123 @@ public void findByColumnAndInc() {
...
+          .then()
+          .statusCode(200)
+          .body("status.matchedCount", is(1))
+          .body("status.modifiedCount", is(1));
+      ;
+
+      String expected =
+          """
+                { "_id":"update_doc_add_to_set_each",
+                  "nested" : { "array": [1, 2, 3, 4] },
```
Comment (ivansenic) [inline src/test/java/io/stargate/sgv2/jsonapi/api/v1/FindAndUpdateIntegrationTest.java:1617]: maybe remove `1` from original array to assert there is no sorting perfomed

File: src/main/java/io/stargate/sgv2/jsonapi/api/model/command/clause/update/AddToSetOperation.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        for (JsonNode element : toAdd) {
+          modified |= addToSet(array, element);
+        }
+      } else {
+        modified |= addToSet(array, toAdd);
+      }
+    }
+
+    // Every valid update operation modifies document so need just one:
+    return modified;
```
Comment (ivansenic) [inline src/main/java/io/stargate/sgv2/jsonapi/api/model/command/clause/update/AddToSetOperation.java:142]: missing int test with modified zero

File: src/main/java/io/stargate/sgv2/jsonapi/util/JsonUtil.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        }
+        Iterator<JsonNode> it = node1.elements();
+        for (JsonNode value : node2) {
+          if (!equalsOrdered(value, it.next())) {
+            return false;
+          }
+        }
+        return true;
+      }
+      case OBJECT -> {
```
Comment (ivansenic) [inline src/main/java/io/stargate/sgv2/jsonapi/util/JsonUtil.java:37]: very strange, you would add two same objects that have fields in different order.. I guess mongo does that, but its very strange

File: src/test/java/io/stargate/sgv2/jsonapi/service/operation/model/command/clause/update/AddToSetOperationTest.java [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+          objectFromJson(
+              """
+              { "x" : 1, "array" : [ [ 1, 2 ], [ 3 ] ] }
+              """);
+      assertThat(doc).isEqualTo(expected);
+    }
+  }
+
+  @Nested
+  @TestMethodOrder(MethodOrderer.OrderAnnotation.class)
```
Comment (ivansenic) [inline src/test/java/io/stargate/sgv2/jsonapi/service/operation/model/command/clause/update/AddToSetOperationTest.java:359]: you dont need this

File: src/test/java/io/stargate/sgv2/jsonapi/service/operation/model/command/clause/update/AddToSetOperationTest.java [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+              """
+              { "array": [ true ], "subdoc" : { "newArray": [ "value" ] } }
+              """);
+      assertThat(doc).isEqualTo(expected);
+    }
+  }
+
+  // Since equality semantics of sub-docs differ, have separate sets for them
+  @Nested
+  @TestMethodOrder(MethodOrderer.OrderAnnotation.class)
```
Comment (ivansenic) [inline src/test/java/io/stargate/sgv2/jsonapi/service/operation/model/command/clause/update/AddToSetOperationTest.java:111]: remove

File: src/test/java/io/stargate/sgv2/jsonapi/service/operation/model/command/clause/update/AddToSetOperationTest.java [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+import org.junit.jupiter.api.MethodOrderer;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestMethodOrder;
+
+@QuarkusTest
+@TestProfile(NoGlobalResourcesTestProfile.Impl.class)
+public class AddToSetOperationTest extends UpdateOperationTestBase {
+  @Nested
+  @TestMethodOrder(MethodOrderer.OrderAnnotation.class)
```
Comment (ivansenic) [inline src/test/java/io/stargate/sgv2/jsonapi/service/operation/model/command/clause/update/AddToSetOperationTest.java:24]: remove

**Review by ivansenic — APPROVED** [review]
Looks good imo..

File: src/main/java/io/stargate/sgv2/jsonapi/util/JsonUtil.java (hunk 4 above)
Comment (tatu-at-datastax) [reply src/main/java/io/stargate/sgv2/jsonapi/util/JsonUtil.java:37]: @ivansenic I agree, it is weird. But it is Mongo semantics :-(

File: src/test/java/io/stargate/sgv2/jsonapi/api/v1/FindAndUpdateIntegrationTest.java (hunk 2 above)
Comment (tatu-at-datastax) [reply src/test/java/io/stargate/sgv2/jsonapi/api/v1/FindAndUpdateIntegrationTest.java:1617]: Good idea, will change. Thanks!

File: src/main/java/io/stargate/sgv2/jsonapi/api/model/command/clause/update/AddToSetOperation.java (hunk 3 above)
Comment (tatu-at-datastax) [reply src/main/java/io/stargate/sgv2/jsonapi/api/model/command/clause/update/AddToSetOperation.java:141]: Ok yes good idea; although unit test covers part of it, need to verify everything is linked. Will add.

File: src/main/java/io/stargate/sgv2/jsonapi/api/model/command/clause/update/AddToSetOperation.java [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+
+    Iterator<Map.Entry<String, JsonNode>> it = actionDef.fields();
+    while (it.hasNext()) {
+      Map.Entry<String, JsonNode> entry = it.next();
+      final String modifier = entry.getKey();
+      final JsonNode arg = entry.getValue();
+
+      switch (modifier) {
+        case "$each":
+          eachArg = arg;
```
[the remark this answers is not in the document]
Comment (tatu-at-datastax) [reply src/main/java/io/stargate/sgv2/jsonapi/api/model/command/clause/update/AddToSetOperation.java:65]: No; neither BSON nor JSON allow duplicate keys for an Object value.

**Review by vkarpov15 — COMMENTED** [review]
LGTM :+1:
